pkg/i2gw/ingress2gateway.go

```diff
@@ -35,70 +35,135 @@ import (
+// be used.
+type Provider interface {
+	CustomResourceReader
+	ResourceConverter
+}
+
+type CustomResourceReader interface {
+
+	// ReadResourcesFromCluster reads custom resources associated with
+	// the underlying Provider implementation from the kubernetes cluster.
+	ReadResourcesFromCluster(ctx context.Context, customResources interface{}) error
+
+	// ReadResourcesFromFiles reads custom resources associated with
+	// the underlying Provider implementation from the files.
+	ReadResourcesFromFiles(ctx context.Context, customResources interface{}, filename string) error
+}
+
+// The ResourceConverter interface specifies all the implemented Gateway API resource
+// conversion functions.
+type ResourceConverter interface {
+
+	// ConvertHTTPRoutes converts the received ingresses and custom resources
+	// associated with the Provider into HTTPRoutes and Gateways.
+	ConvertHTTPRoutes(ingresses []networkingv1.Ingress, customResources interface{}) ([]gatewayv1beta1.HTTPRoute, []gatewayv1beta1.Gateway, field.ErrorList)
+}
+
+func Run(printer printers.ResourcePrinter, namespace string, inputFile string) error {
 	conf, err := config.GetConfig()
 	if err != nil {
-		fmt.Println("failed to get client config")
-		os.Exit(1)
+		return fmt.Errorf("failed to get client config: %w", err)
 	}
 
 	cl, err := client.New(conf, client.Options{})
 	if err != nil {
-		fmt.Println("failed to create client")
-		os.Exit(1)
+		return fmt.Errorf("failed to create client: %w", err)
 	}
 	cl = client.NewNamespacedClient(cl, namespace)
 
-	ingressList := &networkingv1.IngressList{}
-
+	var ingresses networkingv1.IngressList
 	if inputFile != "" {
-		err = constructIngressesFromFile(ingressList, inputFile)
-		if err != nil {
-			fmt.Printf("failed to open input file: %v\n", err)
-			os.Exit(1)
+		if err = constructIngressesFromFile(&ingresses, inputFile); err != nil {
+			return fmt.Errorf("failed to open input file: %w", err)
 		}
 	} else {
-		err = cl.List(context.Background(), ingressList)
-		if err != nil {
-			fmt.Printf("failed to list ingresses: %v\n", err)
-			os.Exit(1)
+		if err = cl.List(context.Background(), &ingresses); err != nil {
+			return fmt.Errorf("failed to list ingresses: %w", err)
 		}
 	}
 
-	if len(ingressList.Items) == 0 {
+	if len(ingresses.Items) == 0 {
 		msg := "No resources found"
 		if namespace != "" {
 			fmt.Printf("%s in %s namespace\n", msg, namespace)
 		} else {
 			fmt.Println(msg)
 		}
-		return
+		return nil
 	}
 
-	httpRoutes, gateways, errors := ingresses2GatewaysAndHTTPRoutes(ingressList.Items)
-	if len(errors) > 0 {
-		fmt.Printf("# Encountered %d errors\n", len(errors))
-		for _, err = range errors {
+	var gateways []gatewayv1beta1.Gateway
+	var httpRoutes []gatewayv1beta1.HTTPRoute
+	var errs field.ErrorList
+
+	providerByName := constructProviders(&ProviderConf{})
+	if len(providerByName) == 0 {
+		return fmt.Errorf("no providers")
+	}
+
+	for name, provider := range providerByName {
```

Comment:
makes sense for now, but as we add providers I would guess we want to add a flag to give users the opportunity to select a specific one? another TODO/issue for that?

Reply:
There is a TODO for this one ;)
Take a look at https://github.com/kubernetes-sigs/ingress2gateway/blob/781a9d9a8858ef0d81f17d10839698fffce04e44/pkg/i2gw/ingress2gateway.go#L158